src/approximate-qvm.lisp

```diff
@@ -0,0 +1,183 @@
+(in-package #:qvm)
+
+(defclass approx-qvm (kraus-qvm)
+  ((t1
+    :initarg :t1
+    :accessor t1
+    :initform nil ; default value
+    :documentation "t1 value to simulate")
+   (t2
+    :initarg :t2
+    :accessor t2
+    :initform nil ; default value
+    :documentation "t2 value to simulate")
+   (kraus-ops
+    :initarg :kraus-ops
+    :accessor kraus-ops
+    :initform (make-hash-table :test 'eql)
+    :documentation "list of lists of kraus operators : ((t1 kraus ops) (t2 kraus ops) ...)"
+    )
+   (fros
+    :initarg :fros
+    :accessor fros
+    :initform (make-hash-table :test 'eql)
+    :documentation "noisy readout encoded as a hash table of qubit to avg readout fidelity")
+   (avg-gate-time
+    :initarg :avg-gate-time
+    :accessor avg-gate-time
+    :initform nil ;(error ":AVG-GATE-TIME is a required initarg ~ to APPROX-QVM")
+    :documentation "the average gate time for a gate in this qvm")
+   ))
+
+
+(defmethod initialize-instance :after ((qvm approx-qvm) &rest args)
+  "This function populates the kraus-ops slot. It checks for any set noise params (T1, T2, etc)
+   and adds the kraus operators representing that noise to the kraus-ops slot. "
+  (declare (ignore args)) 
+  (format t "approx-qvm~%")
+  (if (t1 qvm) 
+      (setf (gethash "t1" (kraus-ops qvm))
+            (generate-damping-kraus-ops (t1 qvm) (avg-gate-time qvm))))
+  (if (t2 qvm) 
+      (setf (gethash "t2" (kraus-ops qvm)) 
+            (generate-damping-dephasing-kraus-ops (t2 qvm) (avg-gate-time qvm))))
+  
+  ;; readout fidelities are the average of the assignment probabilities p(0|0) and p(1|1). 
+  ;; We approximate the readout povm as p(0|0) = p(1|1) = fR0, and 0(0|1) = p(1|0) = 1 - fR0.
+  (if (fros qvm)
+      (maphash #'(lambda (k v) (setf (gethash k (readout-povms qvm)) 
+                                     (list v (- 1.0d0 v) (- 1.0d0 v) v)))
+               (fros qvm))))
+
+
+(defmethod set-gate-time ((qvm approx-qvm) gate-time)
+  (setf (avg-gate-time qvm) gate-time)
+  (set-t1 qvm (t1 qvm))
+  (set-t2 qvm (t2 qvm))
+  )
+
+
+(defmethod set-t1 ((qvm approx-qvm) t1)
+  (setf (t1 qvm) t1)
+  (setf (gethash "t1" (kraus-ops qvm)) (generate-damping-kraus-ops (t1 qvm) (avg-gate-time qvm)))
+  )
+
+(defmethod set-t2 ((qvm approx-qvm) t2)
+    (setf (t2 qvm) t2)
+    (setf (gethash "t2" (kraus-ops qvm)) (generate-damping-dephasing-kraus-ops (t2 qvm) (avg-gate-time qvm)))
+
+  )
+
+(defmethod set-fro ((qvm approx-qvm) qubit fro)
+    (setf (gethash qubit (fros qvm)) fro)
+    (setf (gethash qubit (readout-povms qvm)) (list fro (- 1.0d0 fro) (- 1.0d0 fro) fro))
+  )
+
+
+(defmethod apply-all-kraus-ops ((qvm approx-qvm) (instr quil:gate-application))
+  "Apply all the kraus operators (for all of the noise sources) to the state of the program"
+  
+  (loop for kops being each hash-value of (kraus-ops qvm)
```

Comment:
A nitty nit pick. We follow the style of prefixing `loop` keywords with a `:`. e.g. `(loop :for i :below 10 :collect i)`